src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp

```diff
@@ -33,48 +33,103 @@
 // appleseed.renderer headers.
 #include "renderer/global/globallogger.h"
 #include "renderer/kernel/aov/aovaccumulator.h"
+#include "renderer/kernel/aov/tilestack.h"
+#include "renderer/modeling/frame/frame.h"
 
 // appleseed.foundation headers.
+#include "foundation/image/color.h"
+#include "foundation/image/colorspace.h"
+#include "foundation/image/tile.h"
 #include "foundation/platform/types.h"
 
-using namespace foundation;
-
 namespace renderer
 {
 
 //
 // PixelRendererBase class implementation.
 //
 
-PixelRendererBase::PixelRendererBase()
-  : m_invalid_pixel_count(0)
+const foundation::uint8 InvalidSampleHint = 1;
+const foundation::uint8 CorrectSampleHint = 2;
+
+PixelRendererBase::PixelRendererBase(
+    const Frame&        frame,
+    const size_t        thread_index,
+    const ParamArray&   params)
+  : m_base_params(params)
+  , m_invalid_pixel_count(0)
+  , m_invalid_sample_aov_index(~0)
 {
+    if (m_base_params.m_diagnostics)
+    {
+        m_invalid_sample_aov_index = frame.create_extra_aov_image("invalid_samples");
+
+        if ((thread_index == 0) && (m_invalid_sample_aov_index == size_t(~0)))
+        {
+            RENDERER_LOG_WARNING(
+                "could not create invalid samples aov, maximum number of aovs (" FMT_SIZE_T ") reached.",
+                MaxAOVCount);
+        }
+    }
 }
 
 void PixelRendererBase::on_tile_begin(
-    const Frame&    frame,
-    Tile&           tile,
-    TileStack&      aov_tiles)
+    const Frame&                frame,
+    foundation::Tile&           tile,
+    TileStack&                  aov_tiles)
 {
+    if (m_invalid_sample_aov_index != size_t(~0))
+        m_invalid_sample_diagnostic.reset(
+            new foundation::Tile(tile.get_width(), tile.get_height(), 1, foundation::PixelFormatUInt8));
 }
 
 void PixelRendererBase::on_tile_end(
-    const Frame&    frame,
-    Tile&           tile,
-    TileStack&      aov_tiles)
+    const Frame&                frame,
+    foundation::Tile&           tile,
+    TileStack&                  aov_tiles)
 {
+    if (m_invalid_sample_aov_index != size_t(~0))
+    {
+        const size_t width = tile.get_width();
+        const size_t height = tile.get_height();
+
+        for (size_t y = 0; y < height; ++y)
+        {
+            for (size_t x = 0; x < width; ++x)
+            {
+                foundation::Color<foundation::uint8, 1> sample_state;
+                m_invalid_sample_diagnostic->get_pixel(x, y, sample_state);
+
+                foundation::Color4f color(1.0f, 0.0f, 1.0f, 1.0f);
+
+                if (sample_state[0] == CorrectSampleHint)
+                {
+                    tile.get_pixel(x, y, color);
+                    // 20% of luminance
+                    color.rgb().set(foundation::luminance(color.rgb()));
+                    color.rgb().set(color.r * 0.2f);
```

Comment:
That looks a bit weird. Why not simply

    color.rgb().set(0.2f * luminance(color.rgb()));